src/js/angular/resources/controllers.js

```diff
@@ -65,211 +66,215 @@ resourcesCtrl.controller('ResourcesCtrl', ['$scope', '$timeout', 'MonitoringRest
-                        chart.addData(timestamp, data);
+            return MonitoringRestService.monitorStructures(activeRepository)
+                .then(function(response) {
+                    processResponse(response, (timestamp, data) => {
+                        Object.values($scope.structuresMonitorData).forEach((chart) => {
+                            chart.addData(timestamp, data);
+                        });
                     });
+                });
+        };
+        const getClusterMonitorData = function() {
+            return MonitoringRestService.monitorCluster()
+                .then(function(response) {
+                    processResponse(response, (timestamp, data) => $scope.clusterHealthChart.addData(timestamp, data));
+                });
+        };
 
-                    if ($scope.firstLoad) {
-                        $scope.firstLoad = false;
-
-                        const timer = $timeout(function () {
-                            $scope.loader = false;
-                        }, 500);
 
-                        $scope.$on('$destroy', function () {
-                            $timeout.cancel(timer);
-                        });
+        // TODO: Organize the following somehow
+        const getData = (monitor) => {
+            if (hasMonitorError(monitor.error)) {
+                if (monitor.error.retries === MAX_RETRIES) {
+                    if (monitor.poll) {
+                        $timeout.cancel(monitor.poll);
                     }
+                    return;
                 }
-                structuresMonitorPoll = $timeout(getStructuresMonitorData, POLLING_INTERVAL);
-            }).catch(function (data) {
-                $scope.error = getError(data);
-                $scope.loader = false;
-            });
+                monitor.error.retries++;
+            }
+            return monitor.fetchFn()
+                .then(() => {
+                    clearError(monitor.error);
+                })
+                .catch((error) => {
+                    setError(monitor.error, error);
+                })
+                .finally(() => {
+                    if (monitor.poll) {
+                        $timeout.cancel(monitor.poll);
+                    }
+                    monitor.poll = $timeout(() => getData(monitor), POLLING_INTERVAL);
+                });
         };
-        let clusterMonitorPoll;
-        const getClusterMonitorData = function () {
-            if ($scope.error) {
-                return;
+        const processResponse = (response, dataSetter) => {
+            const data = response.data;
+            if (data) {
+                setChartData(data, dataSetter);
+
+                if ($scope.firstLoad) {
+                    $scope.firstLoad = false;
+
+                    const timer = $timeout(function() {
+                        $scope.loader = false;
+                    }, 500);
+
+                    $scope.$on('$destroy', function() {
```

Comment:
There are 3 destroy callbacks registered in this controller. I think I know why you did it this way, but it's kind of messy. IMO better register these timers somewhere in a higher scope and clear them at once in in a single destroy.

Reply:
I will simplify